### PR TITLE
feat: update OCICluster webhook validations

### LINKED ISSUE #70: Add more validation for OCICluster
**What would you like to be added**:
Now that #55 has been merged we need to work on adding more validation checks for things like the following:

- NetworkSpec (valid CIDRs)
- Region is valid (might need to call to OCI for validation or the OCI SDK might have pre-flight validation we could use)
- ClusterName (length?)
- For cluster update `region` is considered an immutable field. We will want to handle other immutable fields

**Why is this needed**:
Now that #55 is merged we should expand upon our validation. This issue isn't to 100% cover all possibilities, but rather cover what we can now that make sense.

### PR DESCRIPTION
**Pull request by joekr**
**What this PR does / why we need it**:
This expands on the OCICluster webhook adding some more validation during cluster creation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #70

### CODE REVIEW

File: api/v1beta1/validator.go
Comment (joekr) [inline api/v1beta1/validator.go:34]: I picked 150 to give us room to expand if needed.

**Comment by joekr** [comment]
# unit
```
?       github.com/oracle/cluster-api-provider-oci      [no test files]
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta1  0.508s  coverage: 16.1% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/config [no test files]
ok      github.com/oracle/cluster-api-provider-oci/cloud/ociutil        0.505s  coverage: 30.3% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/scope  175.925s        coverage: 81.9% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks    [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn  [no test files]
ok      github.com/oracle/cluster-api-provider-oci/controllers  26.544s coverage: 71.0% of statements
```

# e2e
```
Ran 5 of 21 Specs in 1985.456 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 16 Skipped
```

File: api/v1beta1/ocicluster_webhook.go
Comment (shyamradhakrishnan) [inline api/v1beta1/ocicluster_webhook.go:96]: should we validate compartment cannot chaneg during update?

File: api/v1beta1/validator.go
Comment (shyamradhakrishnan) [inline api/v1beta1/validator.go:57]: region can be blnak right? we should return true if it is blank right?

File: api/v1beta1/validator.go
Comment (shyamradhakrishnan) [inline api/v1beta1/validator.go:64]: I dont think we should validate cluster name, that is a kubernetes artifact name, and I think we should allow whatever Kubernetes allows when you create. resource. So best leave it to kubernetes?

File: api/v1beta1/validator.go
Comment (shyamradhakrishnan) [inline api/v1beta1/validator.go:108]: this is great

File: api/v1beta1/validator.go
Comment (shyamradhakrishnan) [inline api/v1beta1/validator.go:162]: should we check if the subnet role is in one of the enums we have?

File: api/v1beta1/validator.go
Comment (shyamradhakrishnan) [inline api/v1beta1/validator.go:115]: should we add validation that in the provided NSG sec list rules, the cidr is correct?

File: api/v1beta1/ocicluster_webhook.go
Comment (shyamradhakrishnan) [inline api/v1beta1/ocicluster_webhook.go:136]: I cannot gather this, but we should call validate of network spec in update as well I think.

File: api/v1beta1/validator.go
Comment (joekr) [reply api/v1beta1/validator.go:57]: In our case a region can be blank. But a valid region can't be blank. 

So the validate function, to me, seems correct. But in the webhook we don't call it if the region is empty. 

I could see it either way it would be less code if we allowed for blank in the validate function 🤷‍♂️

Comment (shyamradhakrishnan) [reply api/v1beta1/validator.go:57]: hmm yeah, I think the fact that stringRegion == "" check is there seems confusing to me. Can we just do this in validRegions
if blank{
return true
}
// check for invalid region in case of non empty
Since we are not validating much in region anyways, I am ok removing this validation itself?

Comment (joekr) [reply api/v1beta1/validator.go:57]: oh yeah I need to rename `blankRegex` 🤦 
it should just be something like `validnameRegex` or something. Right now it is only checking for spaces in strings. I will update this.

File: api/v1beta1/validator.go
Comment (joekr) [reply api/v1beta1/validator.go:64]: I will pull this out since right now we support 255 characters and Kubernetes cluster names max is 253 characters

File: api/v1beta1/ocicluster_webhook.go
Comment (shyamradhakrishnan) [inline api/v1beta1/ocicluster_webhook.go:117]: log?

**Review by shyamradhakrishnan — COMMENTED** [review]
loog good, should we use logger rather than fmt.Println?

File: api/v1beta1/ocicluster_webhook.go
Comment (joekr) [reply api/v1beta1/ocicluster_webhook.go:117]: yeah the PR is still in draft 😄

**Review by shyamradhakrishnan — APPROVED** [review]
Please run the PR blocking e2e tests before merging.

**Comment by joekr** [comment]
Updated

# unit
```
?       github.com/oracle/cluster-api-provider-oci      [no test files]
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta1  0.250s  coverage: 19.0% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/config [no test files]
ok      github.com/oracle/cluster-api-provider-oci/cloud/ociutil        0.481s  coverage: 27.8% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/scope  175.980s        coverage: 81.9% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks    [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn  [no test files]
ok      github.com/oracle/cluster-api-provider-oci/controllers  26.334s coverage: 71.0% of statements
```

#e2e
```
Ran 5 of 21 Specs in 1827.671 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 16 Skipped
```

**Pull request merged**
